src/scheduler.cpp

```diff
@@ -9,13 +9,14 @@
 #include <assert.h>
 #include <utility>
 
-CScheduler::CScheduler() : nThreadsServicingQueue(0), stopRequested(false), stopWhenEmpty(false)
+CScheduler::CScheduler()
 {
 }
 
 CScheduler::~CScheduler()
 {
     assert(nThreadsServicingQueue == 0);
+    if (stopWhenEmpty) assert(taskQueue.empty());
```

Comment:
In that case, should `stop()` also `assert(nThreadsServicingQueue > 0)` at least when `drain == true`?

Reply:
I agree. Maybe relax the assert to `assert(nThreadsServicingQueue > 0 || taskQueue.empty())` because otherwise `CScheduler c; c.stop();` would assert.

Reply:
I don't think we can do that. The scheduler accepts events before any service thread has been started, also we can ask it to stop before any service threads have been started (they might be started later on).

I will leave it as is for now.

Reply:
Makes sense